### PR TITLE
Support dynamic UI menu query

### PR DESCRIPTION
**Pull request by mrproliu**
Add the feature of dynamic UI menu query, similar to the UI Template feature. By predefining the UI menu, it's automatically written into the database when the system starts. Associate menu items with Dashboard ID, **when there is data in the Service dimension of the Layer corresponding to the Dashboard, the menu is deemed to need to be displayed.**

- [ ] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #<issue number>.
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/docs/en/changes/changes.md).

### CODE REVIEW

File: docs/en/setup/backend/configuration-vocabulary.md [hunk 1]
```diff
@@ -48,6 +48,7 @@ The Configuration Vocabulary lists all available configurations provided by `app
...
+| -                       | -             | uiMenuFetchInterval                                                                                                                                                      | The UI menu should activate fetch interval(in seconds).                                                                                                                                                                                                                                                                                                                                                                                                                    | -                                                     | 20                                                                                           |
```
Comment (wu-sheng) [inline docs/en/setup/backend/configuration-vocabulary.md:51]:
```suggestion
| -                       | -             | uiMenuRefreshInterval                                                                                                                                                      | The period(in seconds) of refreshing the status of all UI menu items.                                                                                                                                                                                                                                                                                                                                                                                                                    | -                                                     | 20                                                                                           |
```

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/menu.yaml [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+# menus: the menus of the left navigation bar
+#   - name: the name of the menu item
+#     icon: icon name
+#     path: link path
+#     dashboard: related dashboard name, the menu item should be activated when the dashboard value is empty or dashboard related scope has services
+#     menus: sub menus of the current item
+
+menus:
+  - name: General Service
+    icon: chart
```
Comment (wu-sheng) [inline oap-server/server-starter/src/main/resources/ui-initialized-templates/menu.yaml:25]: Where are the definitions of these icons? Should UI implement this part?

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/menu.yaml [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    menus:
+      - name: Services
+        path: /general
+        dashboard: General-Root
```
Comment (wu-sheng) [inline oap-server/server-starter/src/main/resources/ui-initialized-templates/menu.yaml:29]: Are you going to use the dashboard name? So, the dashboard ID would be irrelevant

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/menu.yaml (hunk 2 above)
Comment (mrproliu) [reply oap-server/server-starter/src/main/resources/ui-initialized-templates/menu.yaml:25]: Yes, these value is the current dashboard icon name.

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/menu.yaml (hunk 3 above)
Comment (mrproliu) [reply oap-server/server-starter/src/main/resources/ui-initialized-templates/menu.yaml:29]: It's dashboard ID, I needs to using the dashboard ID to query the dashboard configuration.

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/menu.yaml (hunk 2 above)
Comment (wu-sheng) [reply oap-server/server-starter/src/main/resources/ui-initialized-templates/menu.yaml:25]: They seem in low readability. Please rename them. @Fine0830 Please reflect the UI icon names to adopt this PR change.

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/menu.yaml (hunk 3 above)
Comment (wu-sheng) [reply oap-server/server-starter/src/main/resources/ui-initialized-templates/menu.yaml:29]: If that is the ID, let's name it as ID.

Comment (wu-sheng) [reply oap-server/server-starter/src/main/resources/ui-initialized-templates/menu.yaml:29]: This doc should be updated as well, https://skywalking.apache.org/docs/main/next/en/ui/readme/. Add menu change introduction.

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/menu.yaml (hunk 2 above)
Comment (mrproliu) [reply oap-server/server-starter/src/main/resources/ui-initialized-templates/menu.yaml:25]: So, I need to wait for @Fine0830 updates, then I update the names? If It's not defined there, the UI doesn't know what kind of icon should be showing.

Comment (wu-sheng) [reply oap-server/server-starter/src/main/resources/ui-initialized-templates/menu.yaml:25]: You should update first. UI follows only.

File: docs/en/ui/README.md [hunk 4]
```diff
@@ -8,6 +8,41 @@ The left side menu lists all available supported stacks with default dashboards.
 
 Follow the `Official Dashboards` menu to explore all default dashboards on their ways to monitor different tech stacks.
 
+## Sidebar Menu and Marketplace
+
+All available feature menu items are only listed in the marketplace(since 9.6.0). They are only visible on the Sidebar Menu when there are relative services
+being observed by various supported observation agents, such as installed language agents, service mesh platform, OTEL integration.
+
+The menu items defined in `ui-initialized-templates/menu.yaml` are the universal marketplace for all default-supported integration.
+The menu definition supports one and two levels items. The leave menu item should have the `path` and `dashboardID` for navigation, the icon of the level one item is defined from UI side.
```
Comment (wu-sheng) [inline docs/en/ui/README.md:17]: @Fine0830 Please check this, whether `path` is necessary?

Comment (wu-sheng) [reply docs/en/ui/README.md:17]: Why doesn't the level one(browser) menu need this, but level two does?

**Comment by wu-sheng** [comment]
@mrproliu Please recheck the bot scanning results.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/management/ui/menu/UIMenu.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+@Stream(name = UIMenu.INDEX_NAME, scopeId = UI_MENU, builder = UIMenu.Builder.class, processor = ManagementStreamProcessor.class)
+@EqualsAndHashCode(of = "menuId", callSuper = false)
+public class UIMenu extends ManagementData {
+    public static final String INDEX_NAME = "ui_menu";
+    public static final String MENU_ID = "menu_id";
+    public static final String CONFIGURATION = "configuration";
+    public static final String UPDATE_TIME = "update_time";
+
+    @Column(name = MENU_ID)
+    private String menuId;
```
[the remark this answers is not in the document]
Comment (mrproliu) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/management/ui/menu/UIMenu.java:48]: @sonatype-lift ignoreall

File: docs/en/ui/README.md (hunk 4 above)
Comment (Fine0830) [reply docs/en/ui/README.md:17]: It seems that the `name` is necessary, and others fields are not necessary.

File: docs/en/ui/README.md [hunk 6]
```diff
@@ -8,6 +8,41 @@ The left side menu lists all available supported stacks with default dashboards.
...
+menus:
+  - name: General Service
+    icon: general_service
+    menus:
+      - name: Services
+        path: /general
+        dashboardID: General-Root
+      - name: Visual Database
+        path: /database
+        dashboardID: Virtual-Database-Root
```
Comment (wu-sheng) [inline docs/en/ui/README.md:29]: According to the discussion between @Fine0830 and me. We need to change this.
`name`, `icon`, and `layer` are required, others are not.
This is the runtime menu context.
image

path should be generated by UI side directly by `lower case` the `name`, and use `_` replacing the blank between words.

Also, this means the query protocol should be fixed as well, no `path`, but need to add `layer`.
https://github.com/apache/skywalking-query-protocol/blob/1f80cd0af8bc0baf71bb68961a894c110c9d9b79/ui-configuration.graphqls#L32

Comment (Fine0830) [reply docs/en/ui/README.md:29]: I think the `name` should not contain special characters, including spaces.

Comment (mrproliu) [reply docs/en/ui/README.md:29]: > I think the `name` should not contain special characters, including spaces.

So, in the 'Control Plane' of 'Service Mesh', the menu should be 'ControlPlane' and 'ServiceMesh'? Am I right?

Comment (Fine0830) [reply docs/en/ui/README.md:29]: Right.

Comment (wu-sheng) [reply docs/en/ui/README.md:29]: > I think the `name` should not contain special characters, including spaces.

Then how the name is showing? We need `Service Mesh` text to display, that is my meaning of the name.

Comment (wu-sheng) [reply docs/en/ui/README.md:29]: > When menu item is alerting or dashboard, the value of layer should be empty?

Alerting and all menus are not configurable are excluded from this. Those would be hard coded only from UI side.

Comment (mrproliu) [reply docs/en/ui/README.md:29]: > > When menu item is alerting or dashboard, the value of layer should be empty?
> 
> 
> 
> Alerting and all menus are not configurable are excluded from this. Those would be hard coded only from UI side.

Got it. It should be including dashboard settings, alert and settings.

Comment (Fine0830) [reply docs/en/ui/README.md:29]: > > I think the `name` should not contain special characters, including spaces.
> 
> Then how the name is showing? We need `Service Mesh` text to display, that is my meaning of the name.

In the graph, the `name` means ID.  `title` could display on the page.

File: docs/en/ui/README.md [hunk 7]
```diff
@@ -8,6 +8,41 @@ The left side menu lists all available supported stacks with default dashboards.
 
 Follow the `Official Dashboards` menu to explore all default dashboards on their ways to monitor different tech stacks.
 
+## Sidebar Menu and Marketplace
+
+All available feature menu items are only listed in the marketplace(since 9.6.0). They are only visible on the Sidebar Menu when there are relative services
+being observed by various supported observation agents, such as installed language agents, service mesh platform, OTEL integration.
+
+The menu items defined in `ui-initialized-templates/menu.yaml` are the universal marketplace for all default-supported integration.
+The menu definition supports one and two levels items. The leave menu item should have the `layer` for navigation.
```
Comment (wu-sheng) [inline docs/en/ui/README.md:17]:
```suggestion
The menu definition supports one and two levels items. The leaf menu item should have the `layer` for navigation.
```

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/menu.yaml [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+# limitations under the License.
+
+# menus: the menus of the left navigation bar
+#   - title: the title of the menu item
+#     icon: icon name
+#     layer: related layer name, the menu item should be activated when layer has services
+#     menus: sub menus of the current item
+
+menus:
+  - title: GeneralService
```
Comment (wu-sheng) [inline oap-server/server-starter/src/main/resources/ui-initialized-templates/menu.yaml:23]: Title is for display, it should not be like this. `General Service` is proper value.

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/menu.yaml [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+# menus: the menus of the left navigation bar
+#   - title: the title of the menu item
+#     icon: icon name
+#     layer: related layer name, the menu item should be activated when layer has services
+#     menus: sub menus of the current item
+
+menus:
+  - title: General Service
+    icon: general_service
+    menus:
```
Comment (wu-sheng) [inline oap-server/server-starter/src/main/resources/ui-initialized-templates/menu.yaml:25]: @Fine0830 You need to covert the title to what you need for path and name.
We don't define it on the OAP side

**Review by wu-sheng — APPROVED** [review]

**Pull request merged**